processor/batch-processor/src/run.ts

```diff
@@ -110,66 +137,120 @@ class Processor<B extends BlockBase, S> {
         log.info(`prometheus metrics are served on port ${server.port}`)
     }
 
-    private updateProgressMetrics(chainHeight: number, state: HashAndHeight, time?: bigint): void {
+    private async updateProgressMetrics(state: HashAndHeight, time?: bigint): Promise<void> {
+        let chainHeight = await this.chainHeight.get()
+
         this.metrics.setChainHeight(chainHeight)
         this.metrics.setLastProcessedBlock(state.height)
         let left: number
         let processed: number
         if (this.src.getBlocksCountInRange) {
             left = this.src.getBlocksCountInRange({
                 from: this.metrics.getLastProcessedBlock() + 1,
-                to: this.metrics.getChainHeight()
+                to: this.metrics.getChainHeight(),
             })
-            processed = this.src.getBlocksCountInRange({
-                from: 0,
-                to: this.metrics.getChainHeight()
-            }) - left
+            processed =
+                this.src.getBlocksCountInRange({
+                    from: 0,
+                    to: this.metrics.getChainHeight(),
+                }) - left
         } else {
             left = this.metrics.getChainHeight() - this.metrics.getLastProcessedBlock()
             processed = this.metrics.getLastProcessedBlock()
         }
         this.metrics.updateProgress(processed, left, time)
     }
 
-    private async processBatch(prevHead: HashAndHeight, blocks: B[]): Promise<HashAndHeight> {
-        let chainHeight = await this.chainHeight.get()
+    private async processBatch(prevState: HotDatabaseState, finalizedHead: HashAndHeight, blocks: B[]): Promise<HotDatabaseState> {
+        assert(
+            prevState.height < finalizedHead.height ||
```

Comment:
Finalized head reported by the data source might lie below the previously finalized head at the start of processing.

Reply:
> Finalized head reported by the data source might lie below the previously finalized head at the start of processing.

How is it possible?